### ISSUE TITLE
[React] Focus is not trapped inside mounted floating element if another floating element unmounts

### ISSUE THREAD

**Issue by nygardk** [floating-ui/floating-ui#2607]
Firstly, big thanks for this library. It's immensely useful and meticulously crafted.

**Issue description**

A real world scenario: open a modal through a popover menu and close the popover menu at the same time. The modal doesn't receive focus as expected. However, first `TAB` indeed moves the focus inside the modal.

In more high level terms: when opening a new floating element while another floating element gets closed, the focus is trapped unexpectedly to the reference element of the closing one.

**To Reproduce**

[Please see the sandbox I created](https://codesandbox.io/p/sandbox/holy-rgb-m757r3?file=%2Fsrc%2Findex.css%3A47%2C45)

1. Click the "Open menu" button
2. Click the "Show modal and close menu" button
3. Focus is not moved inside the modal as expected

I tested that the problem appears on Safari, Chrome and Firefox.

You can see the issue also from this screen recording:

https://github.com/floating-ui/floating-ui/assets/2855908/136b7e56-61cf-4e8c-be1c-6d3bbeb2945a

---

I would like to understand if this is a `FloatingFocusManager` configuration issue or not and what would be the best way to fix it.

**Comment by atomiks** [comment]
`returnFocus={false}` on `FloatingFocusManager` if the drawer modal is opened is one workaround (or keep the menu open to allow focus to return)

**Comment by nygardk** [comment]
Hey @atomiks and thank you for looking into it and making the (initial) fix!

I've upgraded the sandbox with the newer version of `floating-ui/react` [here](https://codesandbox.io/p/sandbox/dark-flower-kj489g).

I'm still experiencing one issue: when the modal is closed, focus is not returned to the "Open menu" button as expected.

I think the last focused element before the modal opens should be the "Open menu" button, because the popover returned focus there before the modal opened.

WDYT?

**Comment by atomiks** [comment]
Yeah I did notice that - however I tested out the native dialog and it doesn't return focus to anything if its trigger got removed from the DOM. I tried a few things to handle focus "restoration" after removal, and couldn't come up with a decent solution.

However:

> I think the last focused element before the modal opens should be the "Open menu" button, because the popover returned focus there before the modal opened.

Keeping track of all the last ones seems like a smart solution to this. Basically iterating backwards through the list and finding the most recent one that is still on the DOM.
